src/extension/declarative/types.rs

```diff
@@ -0,0 +1,208 @@
+    bound: TypeDefBoundDeclaration,
+    /// A list of type parameters for this type.
+    ///
+    /// Each element in the list is a 2-element list, where the first element is
+    /// the human-readable name of the type parameter, and the second element is
+    /// the type id.
+    #[serde(default)]
+    #[serde(skip_serializing_if = "crate::utils::is_default")]
+    params: Vec<TypeParamDeclaration>,
+}
+
+impl TypeDeclaration {
+    /// Register this type in the given extension.
+    ///
+    /// Types in the definition will be resolved using the extensions in `scope`
+    /// and the current extension.
+    pub fn register<'ext>(
+        &self,
+        ext: &'ext mut Extension,
+        scope: &ExtensionSet,
+        registry: &ExtensionRegistry,
+    ) -> Result<&'ext TypeDef, ExtensionDeclarationError> {
+        let params = self
+            .params
+            .iter()
+            .map(|param| param.make_type_param(ext, scope, registry))
+            .collect::<Result<Vec<TypeParam>, _>>()?;
+        let type_def = ext.add_type(
+            self.name.clone(),
+            params,
+            self.description.clone(),
+            self.bound.into(),
+        )?;
+        Ok(type_def)
+    }
+}
+
+/// A declarative TypeBound definition.
+///
+/// Equivalent to a [`TypeDefBound`]. Provides human-friendly serialization, using
+/// the full names.
+///
+/// TODO: Support derived bounds
+#[derive(
+    Debug, Copy, Clone, Serialize, Deserialize, Hash, PartialEq, Eq, Default, derive_more::Display,
+)]
+enum TypeDefBoundDeclaration {
+    /// The equality operation is valid on this type.
+    Eq,
+    /// The type can be copied in the program.
+    Copyable,
+    /// No bound on the type.
+    #[default]
+    Any,
+}
+
+impl From<TypeDefBoundDeclaration> for TypeDefBound {
+    fn from(bound: TypeDefBoundDeclaration) -> Self {
+        match bound {
+            TypeDefBoundDeclaration::Eq => Self::Explicit(TypeBound::Eq),
+            TypeDefBoundDeclaration::Copyable => Self::Explicit(TypeBound::Copyable),
+            TypeDefBoundDeclaration::Any => Self::Explicit(TypeBound::Any),
+        }
+    }
+}
+
+/// A declarative type parameter definition.
+///
+/// Serialized as a 2-element list, where the first element is an optional
+/// human-readable name of the type parameter, and the second element is the
+/// type id.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+struct TypeParamDeclaration {
+    /// The name of the parameter. May be `null`.
+    ///
+    /// TODO: This field is ignored, there's no place to put it in a `TypeParam`.
+    description: Option<String>,
+    /// The parameter type.
+    type_name: TypeName,
+}
+
+impl TypeParamDeclaration {
+    /// Create a [`TypeParam`] from this declaration.
+    ///
+    /// Resolves any type ids using both the current extension and any other in `scope`.
+    ///
+    /// TODO: Only non-parametric opaque types are supported for now.
```

Comment:
what's the reason for starting with this? If I were to pick just one I feel like
I'd pick TypeParam::Type to support rather than TypeParam::Opaque. Simpler and
more commonly used no?

Reply:
I went with supporting the first example.

Having any type parameter is currently rejected when translating the declaration, so this was mostly a check to make sure things can be defined correctly.

We'll probably want to support both once we re-enable type params.